src/WrapCalls.cpp

```diff
@@ -0,0 +1,156 @@
+            iter->second = wrapper;
+            return;
+        } else if (wrapper.same_as(iter->first)) {
+            debug(4) << "Merging wrapper of " << in_func.name() << " [" << wrapped_func.name()
+                     << ", " << wrapper.name() << "] with [" << iter->first.name() << ", "
+                     << iter->second.name() << "]\n";
+            wrappers_map.emplace(wrapped_func, iter->second);
+            wrappers_map.erase(iter);
+            return;
+        }
+    }
+    wrappers_map[wrapped_func] = wrapper;
+}
+
+pair<vector<Function>, map<string, Function>> wrap_func_calls(
+            const vector<Function> &outputs, const map<string, Function> &env) {
+    vector<Function> wrapped_outputs;
+    map<string, Function> wrapped_env;
+
+    // Create empty copies of all functions in env.
+    map<Function, Function> copied_map; // Original Func -> Copy
+    for (const auto &iter : env) {
+        copied_map[iter.second] = Function(iter.second.name());
+    }
+
+    // Deep copy the Func in env into its corresponding empty copy.
+    for (const auto &iter : env) {
+        iter.second.deep_copy(copied_map[iter.second], copied_map);
+    }
+
+    // Need to substitute-in the old reference within the Func with the
+    // deep-copy version.
+    for (auto &iter : copied_map) {
+        iter.second.substitute_calls(copied_map);
+    }
+
+    map<Function, map<Function, Function>> func_wrappers_map; // In Func -> [wrapped Func -> wrapper]
+
+    // Populate the env with the deep-copy version.
+    for (const auto &iter : copied_map) {
+        wrapped_env.emplace(iter.first.name(), iter.second);
+        func_wrappers_map[iter.second];
+    }
+
+    for (const auto &it : copied_map) {
+        string wrapped_fname = it.first.name();
+        const Function &wrapped_func = it.second;
+        const auto &wrappers = wrapped_func.schedule().wrappers();
+
+        // Put names of all wrappers of this Function into the set for
+        // faster comparison during the substitution.
+        set<string> all_func_wrappers;
+        for (const auto &iter : wrappers) {
+            all_func_wrappers.insert(Function(iter.second).name());
+        }
+
+        for (const auto &iter : wrappers) {
+            string in_func = iter.first;
+            const Function &wrapper = Function(iter.second); // This is already the deep-copy version
+            internal_assert(is_copy(wrapper, copied_map));   // Make sure it's indeed the copy
+
+            if (in_func.empty()) { // Global wrapper
+                for (const auto &wrapped_env_iter : wrapped_env) {
+                    in_func = wrapped_env_iter.first;
+                    if ((wrapper.name() == in_func) || (all_func_wrappers.find(in_func) != all_func_wrappers.end())) {
+                        // Should not substitute itself or custom wrapper of the same func
+                        debug(4) << "Skip over replacing \"" << in_func << "\" with \"" << wrapper.name() << "\"\n";
```

Comment:
Suggested comment: The only Func that should still call the original is the wrapper itself, so we don't want to rewrite the calls done by the wrapper.